Implement the superchain config contract

maurelian opened the pull request:
**Description**

A rough sketch of the arch of a new `SuperchainConfig` contract.

**TODOs:**

Impl:

- [x] Add getters and setters
- [x] Implement the correct auth for setting each of values in the SuperchainConfig contract
- [x] Ensure storage layout mods in `SystemConfig` are safe

Testing/Docs:

- [x] Get unit tests working
- [x] Update deployment scripts
- [x] Define invariant tests


Open todos: 
- [x] Update config var names with the prefix of the contract they are defined in. 
- [x] add `maxPause` as a config var

maurelian committed:
feat(ctb): Add outline of SuperchainConfig

This commit adds the variables, events, and initializer.

tynes commented on packages/contracts-bedrock/src/L1/OptimismPortal.sol:
```diff
@@ -83,53 +90,46 @@ contract OptimismPortal is Initializable, ResourceMetering, ISemver {
         systemConfig = _systemConfig;
-        guardian = _guardian;
-        paused = _paused;
+        superchainConfig = _superchainConfig;
```
Is the `guardian()(address)` getter still present?

maurelian replied:
'tis.

tynes commented on packages/contracts-bedrock/src/L1/OptimismPortal.sol:
```diff
@@ -148,21 +148,7 @@ contract OptimismPortal is Initializable, ResourceMetering, ISemver {
     /// @notice Getter for the Guardian
     /// @custom:legacy
     function GUARDIAN() external view returns (address) {
```
We can deprecate `GUARDIAN()(address)` i think

tynes commented on packages/contracts-bedrock/src/L1/OptimismPortal.sol:
```diff
@@ -46,10 +47,10 @@ contract OptimismPortal is Initializable, ResourceMetering, ISemver {
-    /// @notice Determines if cross domain messaging is paused.
-    ///         When set to true, withdrawals are paused.
-    ///         This may be removed in the future.
-    bool public paused;
```
Add getter `paused()(bool)` that wraps superchain config

tynes commented on packages/contracts-bedrock/src/L1/SystemConfig.sol:
```diff
@@ -70,6 +73,10 @@ contract SystemConfig is OwnableUpgradeable, ISemver {
+    /// @notice SuperchainConfig address
+    ///         This can only be set via upgrade.
+    SuperchainConfig public superchainConfig;
+
```
needs to go to bottom of list of storage variables

tynes commented on packages/contracts-bedrock/src/l1/SuperchainConfig.sol:
```diff
@@ -0,0 +1,122 @@
+        vetoer = _vetoer;
+        guardian = _guardian;
+        delay = _delay;
+        paused = false;
```
This doesn't use the variable passed in but hardcodes `false`?

maurelian replied:
fixed.

tynes commented on packages/contracts-bedrock/src/l1/SuperchainConfig.sol:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    /// @notice Event version identifier.
+    uint256 public constant VERSION = 0;
```
do we need this?

maurelian replied:
no will remove.

maurelian committed:
refactor(ctb): move Sequencer keys to libs

messing around with tests

maurelian committed:
feat(ctb): SystemConfig ensure sequencer is allowed.

maurelian committed:
Comment out Multichain deploy script

maurelian committed:
refactor(ctb): move pause from Portal to SuperchainConfig

maurelian committed:
feat(ctb): Add removeSequencer()

maurelian committed:
feat(ctb): Add SystemOwner role to SuperchainConfig

maurelian committed:
feat(ctb): Add permissionless removal of an unauthorized sequencer

maurelian committed:
refactor(ctb): Remove legacy GUARDIAN getter.

maurelian committed:
refactor(ctb): Remove VERSION constant and unnecessary spacer

maurelian commented on packages/contracts-bedrock/src/L1/OptimismPortal.sol:
```diff
@@ -83,53 +85,47 @@ contract OptimismPortal is Initializable, ResourceMetering, ISemver {
-    event Unpaused(address account);
-
+    // TODO: move to a new inheritable contract so that this modifier can be reused in the
+    // xDM and bridge
```
This coment is wrong. 
I do still need to expand the pause to the other bridge contracts, but these events don't need to be in them.

maurelian committed:
refactor(ctb): Use eip1967 style storage layout

This approach is consistent with the SystemConfig and ProtocolVersions contracts.

tynes commented on packages/contracts-bedrock/src/L1/OptimismPortal.sol:
```diff
@@ -147,22 +143,14 @@ contract OptimismPortal is Initializable, ResourceMetering, ISemver {
 
     /// @notice Getter for the Guardian
     /// @custom:legacy
```
The lowercase getter is not legacy

tynes replied:
We will want to keep the legacy `GUARDIAN()(address)` getter around unfortunately

maurelian replied:
OK, I can revert that commit, but are you sure? I was addressing [this comment](https://github.com/ethereum-optimism/optimism/pull/7310#discussion_r1331666133).

tynes commented on packages/contracts-bedrock/src/l1/SuperchainConfig.sol:
```diff
@@ -0,0 +1,247 @@
+        emit ConfigUpdate(UpdateType.SYSTEM_OWNER, abi.encode(_systemOwner));
+
+        _setAddress(INITIATOR_SLOT, _initiator);
+        emit ConfigUpdate(UpdateType.INITIATOR, abi.encode(_initiator));
```
The `ConfigUpdate` event in the `ChainConfig` was meant to be consumed by the `op-node`. I don't know how useful the `[type, bytes]` style events are in this context

maurelian replied:
yeah, I'm not attached to this, but I felt like there was a certain "Config Contract" pattern emerging from the `ProtocolVersion` and `SystemConfig` contracts. 

If that's not the case, then IMO we should just get rid of the whole eip1967 layout as well, it brings a ton of boilerplate with it.

maurelian replied:
@tynes thinking about it more, I do like this format, whether or not the op-node reads it. 
Config contracts are inevitably going to get more and more values, having to add a new event for each one is onerous, and doesn't help off-chain code that much. 

Please LMK if you still wnat me to change it.

tynes commented on packages/contracts-bedrock/src/l1/SuperchainConfig.sol:
```diff
@@ -0,0 +1,247 @@
+        reinitializer(2)
+    {
+        _setAddress(SYSTEM_OWNER_SLOT, _systemOwner);
+        emit ConfigUpdate(UpdateType.SYSTEM_OWNER, abi.encode(_systemOwner));
```
Prefer internal functions and not emitting the events directly, will help to ensure the events are always emitted if there are changes in the future

```solidity
function _setSystemOwner(address _systemOwner) internal {
  _setAddress(SYSTEM_OWNER_SLOT, _systemOwner);
  emit ConfigUpdate(...);
}
```

maurelian replied:
Can do, note however that these variables (with the exception of `paused`) are all updated via upgrade, so we'd be creating 6 new setters that are only used in the initializer.

maurelian replied:
Did this.

maurelian committed:
feat(ctb): Move the pause function to SuperchainConfig

The pause() method is moved to the SuperchainConfig contract,
and also applied to the L1xDM's relayMessage() function.

maurelian committed:
feat(ctb): Add SuperchainConfig to deploy script

maurelian committed:
chore: gas, lint, semver, invariant-docs

maurelian committed:
fixup! feat(ctb): SystemConfig ensure sequencer is allowed.

maurelian committed:
fix(ctb): Remove SequencerKeys from SystemConfig interface

If the type is included in two different interfaces, it causes
a naming collision. This seems to be the only way to fix it.

maurelian committed:
refactor(ctb): Add internal setter functions to SuperchainConfig

maurelian committed:
feat(chain-ops): Add SuperchainConfig values to config

chore: Add Superchain to bindings
feat(ctb): Add Superchain config vars to DeployConfig
Update op-e2e tests with new system_config iface
Also increases the evm memory_limit as otherwise the deploy.sync() method will fail with memory OOG.

maurelian committed:
chore: bindings

maurelian committed:
fix(ctb): restore legacy GUARDIAN getter

maurelian committed:
feat(ctb): make pause temporary

maurelian commented on op-chain-ops/genesis/testdata/l1-deployments.json:
```diff
@@ -18,5 +18,7 @@
-  "ProtocolVersionsProxy": "0x3732a4D4Ab006cA4825822baEA1569A107683fa1"
+  "ProtocolVersionsProxy": "0x3732a4D4Ab006cA4825822baEA1569A107683fa1",
+  "SuperchainConfig": "0x883C06a27D76B1CEbdf7EB376f5556c355afC8e5",
+  "SuperchainConfigProxy": "0x3732a4D4Ab006cA4825822baEA1569A107683fa1"
```
These are not the correct values. I just put them in to get the correct address from the failing test.

tynes replied:
This test fixture is for json serialization?

maurelian replied:
Yes. All it does is check that the addresses are non-zero. https://github.com/ethereum-optimism/optimism/blob/develop/op-chain-ops/genesis/testdata/l1-deployments.json#L20-L21

maurelian commented on packages/contracts-bedrock/.storage-layout:
```diff
@@ -68,28 +68,39 @@
 | l2Oracle             | contract L2OutputOracle                                    | 53   | 1      | 20    | src/L1/OptimismPortal.sol:OptimismPortal |
 | systemConfig         | contract SystemConfig                                      | 54   | 0      | 20    | src/L1/OptimismPortal.sol:OptimismPortal |
-| guardian             | address                                                    | 55   | 0      | 20    | src/L1/OptimismPortal.sol:OptimismPortal |
+| superchainConfig     | contract SuperchainConfig                                  | 55   | 0      | 20    | src/L1/OptimismPortal.sol:OptimismPortal |
```
I believe this is a safe overwrite to make.

tynes replied:
I agree this is safe to overwrite this

maurelian commented on packages/contracts-bedrock/foundry.toml:
```diff
@@ -24,6 +24,9 @@ bytecode_hash = 'none'
 build_info_path = 'artifacts/build-info'
 
+# Accomodate the Deploy script's memory intensive sync() method
+memory_limit = 67108864
```
I was getting a `MemoryLimitOOG` error.

maurelian commented on packages/contracts-bedrock/src/L1/L1CrossDomainMessenger.sol:
```diff
@@ -53,4 +53,16 @@ contract L1CrossDomainMessenger is CrossDomainMessenger, ISemver {
+
+    /// @inheritdoc CrossDomainMessenger
+    function paused() public view override returns (bool) {
+        return PORTAL.paused();
```
PORTAL.paused() is [here](https://github.com/ethereum-optimism/optimism/pull/7310/files#diff-f5f7dc90c748b7d98110fa1cc80f3bedd5aa9f299b691c51b9bb473a20b526caR158), it just calls out to the SuperchainConfig to get the paused status. 

IMO this should be changed to also call to the SystemConfig, but this approach made the refactoring easier so far.

tynes replied:
I agree that we should not chain the calls of paused thru the portal, do you plan on updating this?

maurelian replied:
Yes, I noticed this open item this morning.

maurelian commented on packages/contracts-bedrock/src/L1/SuperchainConfig.sol:
```diff
@@ -0,0 +1,270 @@
+
+    // todo(maurelian): make this time configurable. It was just a lot easier to mock it up by hardcoding it in.
+    /// @notice The maximum time in seconds that the system can be paused for.
+    uint256 public constant maxPause = 1 weeks;
```
Need a new config value to make this not constant.

maurelian commented on packages/contracts-bedrock/src/L1/SuperchainConfig.sol:
```diff
@@ -0,0 +1,270 @@
+    /// @notice Emitted when configuration is updated.
+    /// @param updateType Type of update.
+    /// @param data       Encoded update data.
+    event ConfigUpdate(UpdateType indexed updateType, bytes data);
```
I realize that this event format is used in other contracts because the op-node reads one, but I like it here regardless, and don't think it makes sense to have a different event for each config var being changed.

maurelian commented on packages/contracts-bedrock/src/L1/SystemConfig.sol:
```diff
@@ -89,25 +92,30 @@ contract SystemConfig is OwnableUpgradeable, ISemver {
+    /// @notice Semantic version.
+    /// @custom:semver 2.0.0
+    string public constant version = "2.0.0";
+
```
This diff moves the ConfigUpdate event below the variable declarations.
It should not cause issues with the storage layout.

maurelian commented on packages/contracts-bedrock/src/L1/SystemConfig.sol:
```diff
@@ -286,9 +297,32 @@ contract SystemConfig is OwnableUpgradeable, ISemver {
+    /// @param _batcherHash New batch hash.
     /// @param _unsafeBlockSigner New unsafe block signer address.
-    function setUnsafeBlockSigner(address _unsafeBlockSigner) external onlyOwner {
+    function setSequencer(bytes32 _batcherHash, address _unsafeBlockSigner) external onlyOwner {
```
Note that there is now only a single `setSequencer` method, rather than separate `setUnsafeBlockSigner` and `setBatcherHash`. 
This is because they are defined in the `SuperchainConfig` allowlist as key pairs, so it does not make sense to separate them out here.

maurelian commented on packages/contracts-bedrock/src/libraries/Types.sol:
```diff
@@ -67,4 +67,13 @@ library Types {
+    /// @notice Struct representing a sequencer.
+    /// @custom:field batcherHash         Identifier for the batcher.
+    /// @custom:field unsafeBlockSigner   Unsafe block signer address
+    struct SequencerKeys {
```
I'd like to rename this to SequencerKeyPair

tynes replied:
I am on board with this

maurelian replied:
I have done so

maurelian commented on packages/contracts-bedrock/test/CommonTest.t.sol:
```diff
@@ -169,58 +170,113 @@ contract L2OutputOracle_Initializer is CommonTest {
-    OptimismPortal internal opImpl;
-    OptimismPortal internal op;
-    SystemConfig systemConfig;
+contract SuperchainConfig_Initializer is L2OutputOracle_Initializer {
```
This diff is hard to read, what I did is:
1. add a new `SuperchainConfig_Initializer` contract
2. move `SystemConfig_Initializer` to this file rather than in `SystemConfig.t.sol`
3. Put both of those before `Portal_Initializer` because it depends on them being deployed already.

maurelian commented on packages/contracts-bedrock/test/OptimismPortal.t.sol:
```diff
@@ -15,85 +15,30 @@ import { ResourceMetering } from "../src/L1/ResourceMetering.sol";
-        op.unpause();
-
-        assertEq(op.paused(), true);
-    }
```
I moved these tests to `SuperchainConfig.t.sol`

maurelian commented on packages/contracts-bedrock/test/SystemConfig.t.sol:
```diff
@@ -2,76 +2,21 @@
-    }
-}
-
-contract SystemConfig_Initialize_Test is SystemConfig_Init {
```
moved all of this stuff to CommonTest

tynes commented on packages/contracts-bedrock/scripts/Deploy.s.sol:
```diff
@@ -604,11 +644,33 @@ contract Deploy is Deployer {
+                SuperchainConfig.initialize,
+                (
+                    cfg.finalSystemOwner(), // systemOwner
+                    cfg.securityCouncil(), // initiator
```
It feels like a bit of a leaky abstraction to have security council in the deploy config instead of initiator. Something that I'd like to strive for is consistency in the language. We already have some issues with this, like unsafe block signer vs p2p sequencer key. Since initiator is used in the contract, I'd prefer to keep that language in the deploy config. We can add a comment saying that in practice, it is the security council

maurelian replied:
I've set these back to the more standard names

tynes commented on op-chain-ops/genesis/config.go:
```diff
@@ -204,6 +214,15 @@ type DeployConfig struct {
 	RecommendedProtocolVersion params.ProtocolVersion `json:"recommendedProtocolVersion"`
+	// SecurityCouncil represents the address of the Security Council role.
+	SecurityCouncil common.Address `json:"securityCouncil"`
+	// Foundation represents the address of the Founcation role.
```
nit: typo with `founcation`

maurelian commented on packages/contracts-bedrock/src/L1/SystemConfig.sol:
```diff
@@ -286,9 +297,32 @@ contract SystemConfig is OwnableUpgradeable, ISemver {
         _setUnsafeBlockSigner(_unsafeBlockSigner);
+        _setBatcherHash(_batcherHash);
+
+        Types.SequencerKeys memory _sequencer =
+            Types.SequencerKeys({ unsafeBlockSigner: _unsafeBlockSigner, batcherHash: _batcherHash });
```
@tynes Something you may notice is that `setSequencer()` could accept as struct arg, which I think would be cleaner, ie.

```
setSequencer(Types.SequencerKeys memory _sequencer)
```

However when I put it into the interface, the [same type](https://github.com/ethereum-optimism/optimism/pull/7310/files#diff-e463331be751e9c6fa0f44ee6b7534067c0db4728537611f6575b328fe533b26R32-R35) which is currently in `superchainconfig.go` is also added to `systemconfig.go` and I get this error: 

```
./op-bindings/bindings/systemconfig.go:52:6: TypesSequencerKeys redeclared in this block
../op-bindings/bindings/superchainconfig.go:32:6: other declaration of TypesSequencerKeys
```

It's the reason I renamed the import [here](https://github.com/ethereum-optimism/optimism/pull/7310/files#diff-016d692e745a10a098159957d937a3a02c5317f6bf727d24c584f8448e81cfceR6), but that didn't change the name of the struct in the binding. I wonder if you've encountered that before or can think of a workaround?

tynes replied:
I did have a problem similar to this when doing the gas config in the system config/resource metering contracts. I can't remember exactly how I handled it to be honest. It is trying to generate the type twice in the same package, this is a problem with how we namespace our bindings

tynes replied:
Ah i understand, its trying to generate the type that is the input to the function and its not smart enough to know that the type is already generated in a different file

tynes replied:
The exclude functionality on abigen could work, there is also a combined json flag on abigen that we don't use and i wonder if that would help, altho idk how to get the combined json out of forge plus it would prob include the test files and stuff

tynes replied:
My sense for now its something that we just leave as is and then we solve it sometime in the future

maurelian replied:
Would it be too brittle to find and delete the duplicate type def from one of the bindings at the end of generation?

maurelian replied:
Going to resolve this. In a future clean up I might handle this with --exc

tynes commented on packages/contracts-bedrock/scripts/Deploy.s.sol:
```diff
@@ -604,11 +644,33 @@ contract Deploy is Deployer {
+
+        bytes32 batcherHash = bytes32(uint256(uint160(cfg.batchSenderAddress())));
+
+        Proxy(superchainConfigProxy).upgradeToAndCall({
```
This is being done thru the eoa and not thru the safe. If you feel like this is the easiest way to do it then i'm fine but it would be breaking the standard way that other things are initialized

maurelian replied:
Fixed. The `superchainConfigProxy` is now deployed with the Safe as it's admin.

tynes commented on packages/contracts-bedrock/scripts/Deploy.s.sol:
```diff
@@ -62,14 +64,30 @@ contract Deploy is Deployer {
+    }
+
+    /// @notice Deploy a new OP Chain, with an existing SuperchainConfig provided
+    function setupOpChain(address supConf) public {
```
Technically setting up an op chain doesn't need to deploy the implementations

tynes replied:
We don't need to worry about this now and can eventually follow up with a deploy script that has clean abstractions for:
- creating superchain (deploy superchain config)
- deploying implementations
- deploying proxies
- initializing proxies

maurelian replied:
Good call out. I think that likely will need to be resolved before merging back into `develop`. Going to leave this comment open for now.

maurelian replied:
Resolving now.

tynes commented on packages/contracts-bedrock/scripts/Deploy.s.sol:
```diff
@@ -367,6 +395,19 @@ contract Deploy is Deployer {
         addr_ = address(proxy);
     }
 
```
nit: no natspec here

tynes commented on packages/contracts-bedrock/scripts/Deploy.s.sol:
```diff
@@ -604,11 +644,33 @@ contract Deploy is Deployer {
         console.log("DisputeGameFactory version: %s", version);
     }
 
+    function initializeSuperchainConfig() public broadcast {
```
nit: no natspec

tynes commented on packages/contracts-bedrock/scripts/Deploy.s.sol:
```diff
@@ -127,6 +145,16 @@ contract Deploy is Deployer {
 
+    function deploySuperchainConfigProxy() public broadcast returns (address addr_) {
+        Proxy proxy = new Proxy({
+            _admin: msg.sender
```
It looks like the deployer is the admin of the proxy at the end of the script, we want the proxy admin to be the admin?

maurelian replied:
Fixed

tynes commented on packages/contracts-bedrock/src/L1/L1CrossDomainMessenger.sol:
```diff
@@ -53,4 +53,16 @@ contract L1CrossDomainMessenger is CrossDomainMessenger, ISemver {
     function _isUnsafeTarget(address _target) internal view override returns (bool) {
         return _target == address(this) || _target == address(PORTAL);
     }
+
```
We need minor semver bump here

tynes commented on packages/contracts-bedrock/src/L1/SuperchainConfig.sol:
```diff
@@ -0,0 +1,280 @@
+    ///         These storage slots decouple the storage layout from
+    ///         solc's automation.
+    /// @param _slot The storage slot to retrieve the address from.
+    function _getValue(bytes32 _slot) internal view returns (uint256 value_) {
```
Perhaps a follow up would be to turn these sorts of getters into a library because we have them defined in multiple contracts

maurelian replied:
yeah, totally agree.

tynes commented on packages/contracts-bedrock/src/L1/SuperchainConfig.sol:
```diff
@@ -0,0 +1,280 @@
+    /// @notice Getter for the delay address.
+    function delay() public view returns (uint256 delay_) {
+        // We do some casting rather than define a new getter.
+        delay_ = uint256(uint160(_getAddress(DELAY_SLOT)));
```
This seems bug prone if the delay is more than 160 bits, unlikely in practice but imo these sorts of shortcuts create the habits that lead to bugs

maurelian replied:
hmmm, looks like I already have a getValue() I can use there anyways.

maurelian committed:
feat(ctb): Initialize SuperchainConfigProxy via Safe

This removes a call from an EOA, making the script more consistent with other initialization calls.

maurelian committed:
feat(ctb): Remove transferProxyAdminOwnership step from deploy

Instead the ProxyAdmin is created with the system owner safe as its owner,

and initialization is done using _callViaSafe.

maurelian committed:
chore(ctb): Bump semver on L1xDM

maurelian committed:
feat(ctb): Move storage getters and setters to own lib

tynes commented on packages/contracts-bedrock/scripts/DeployConfig.s.sol:
```diff
@@ -50,6 +51,10 @@ contract DeployConfig is Script {
+    address public updateInitiator;
+    address public updateVetoer;
+    uint256 public updateDelay;
+    Types.SequencerKeys[] public sequencerKeys;
```
Did you still want to rename this to sequencer key pair?

tynes commented on packages/contracts-bedrock/src/L1/SuperchainConfig.sol:
```diff
@@ -0,0 +1,238 @@
+
+import { Initializable } from "@openzeppelin/contracts/proxy/utils/Initializable.sol";
+import { ISemver } from "src/universal/ISemver.sol";
+import { Types as T } from "src/libraries/Types.sol";
```
With the import causing bindings issues, in the future we can use the `abigen`  flag for excluding types, its `abigen -exc` but that would require a change to the bindings build script/config. If you would like to take it on to get the preferable API, that is cool but also its all good if not

maurelian replied:
I think we need to aggressively cut scope here, resolving :)

tynes commented on packages/contracts-bedrock/src/L1/SuperchainConfig.sol:
```diff
@@ -0,0 +1,238 @@
+        _setDelay(_delay);
+
+        for (uint256 i = 0; i < _sequencers.length; i++) {
+            bytes32 sequencerHash = Hashing.hashSequencerKeys(_sequencers[i]);
```
Curious if you think this should call `_addSequencer` and then make `addSequencer` have the auth check

maurelian replied:
Yes, I like that, implemented.

tynes commented on packages/contracts-bedrock/src/L1/SuperchainConfig.sol:
```diff
@@ -0,0 +1,238 @@
+
+    /// @notice Getter for the paused address.
+    function paused() public view returns (bool paused_) {
+        paused_ = Slot.getUint(PAUSED_SLOT) > block.timestamp;
```
Should this be `>` or `>=`?

maurelian replied:
I think `>` because that way it is paused until the end up `duration`.

maurelian committed:
chore(ctb): Rename SequencerKeys type to SequencerKeyPair

maurelian committed:
feat(ctb): Add maxPause variable to SuperchainConfig

This commit also standardizes the deploy config variable names used to
have the contract name prefix, ie. superchainConfigMaxPause

tynes commented on op-chain-ops/genesis/config.go:
```diff
@@ -31,6 +31,12 @@ var (
 	ErrInvalidImmutablesConfig = errors.New("invalid immutables config")
 )
 
+// SequencerKeys is an auto generated low-level Go binding around an user-defined struct.
```
Auto generated? I feel like this should say something along the lines of

```
SequencerKeyPair represents the identity of a particular sequencer
```

tynes commented on op-chain-ops/genesis/config.go:
```diff
@@ -204,6 +210,17 @@ type DeployConfig struct {
+	// SequencerKeys represents the set of allowed Sequencer key pairs (batcher hash and unsafe
+	// block signer).
+	SuperchainConfigSequencerKeyPairs []SequencerKeyPair `json:"superchainConfigSequencerKeyPairs"`
 }
```
We should be sure to either log warnings in `Check` or error. If they must be set, then we should error

maurelian replied:
adding that

tynes commented on packages/contracts-bedrock/src/L1/OptimismPortal.sol:
```diff
@@ -83,53 +85,47 @@ contract OptimismPortal is Initializable, ResourceMetering, ISemver {
-    event Unpaused(address account);
-
+    // TODO: move to a new inheritable contract so that this modifier can be reused in the
+    // xDM and bridge
```
Just want to make sure that this todo is removed

maurelian replied:
yeah, not worth adding inheritance IMO.

tynes commented on packages/contracts-bedrock/src/L1/SuperchainConfig.sol:
```diff
@@ -0,0 +1,260 @@
+        paused_ = Slot.getUint(PAUSED_SLOT) > block.timestamp;
+    }
+
+    // todo(maurelian): we might need a repause() getter which is only callable by the
```
Following up on this `pause`/`repause` - what is the thinking here?

maurelian replied:
We do need this, but I would prefer to include it in a follow up PR just to get this closed and so that that can happen in the context of a smaller diff.

tynes commented on packages/contracts-bedrock/src/L1/SystemConfig.sol:
```diff
@@ -167,20 +178,21 @@ contract SystemConfig is OwnableUpgradeable, ISemver {
         _setGasConfig({ _overhead: _overhead, _scalar: _scalar });
         _setGasLimit(_gasLimit);
         _setUnsafeBlockSigner(_unsafeBlockSigner);
 
```
The above setters emit events, the contracts being stored do not need to emit events so it is safe to use `Slot` directly altho i thought it was renamed to `Storage` in another PR?

maurelian replied:
I put that PR on `develop` to minimize drift between it and this branch. Now I'm not sure it's working out that way.

maurelian replied:
I implemented this using the Slot lib.

tynes commented on packages/contracts-bedrock/src/universal/CrossDomainMessenger.sol:
```diff
@@ -219,6 +219,10 @@ abstract contract CrossDomainMessenger is
     {
+        // On L1 this function will check the Portal for its paused status.
+        // On L2 this function should be a no-op.
+        _revertWhenPaused();
```
Is this necessary? Why can't we do:

```solidity
require(paused() == false, "OptimismPortal: paused");
```

maurelian replied:
My thinking was that calling a function the just returns is more likely to be optimized away than this.
I'm not certain either way, might do a little sandbox test, or else just take your suggestion.

maurelian replied:
I verified this.

maurelian committed:
feat(ctb): Make SystemConfig.superchainConfig a 1967 var

maurelian committed:
feat(ctb): L1xDM reads paused status from SystemConfig

maurelian committed:
chore: bindings and semver-lock

tynes commented on packages/contracts-bedrock/src/L1/OptimismPortal.sol:
```diff
@@ -83,53 +85,47 @@ contract OptimismPortal is Initializable, ResourceMetering, ISemver {
-    event Unpaused(address account);
-
+    // TODO: move to a new inheritable contract so that this modifier can be reused in the
+    // xDM and bridge
```
why are you resolving these comments and not removing the todo?

maurelian replied:
it's removed now

maurelian committed:
fix(ctb): Remove todos

tynes commented:
I'd like to merge https://github.com/ethereum-optimism/optimism/pull/7428 and then rebase on top of it

tynes commented on packages/contracts-bedrock/src/L1/SystemConfig.sol:
```diff
@@ -286,26 +280,43 @@ contract SystemConfig is OwnableUpgradeable, ISemver {
+            Types.SequencerKeyPair({ unsafeBlockSigner: unsafeBlockSigner(), batcherHash: batcherHash });
+        bytes32 seqHash = Hashing.hashSequencerKeyPair(sequencer);
+        if (SuperchainConfig(superchainConfig()).allowedSequencers(seqHash)) {
+            revert("SystemConfig: cannot remove allowed sequencer.");
```
Why a revert here with a string instead of a require?

maurelian replied:
I find this framing to be more readable, ie. saying "if it is allowed, then stop the check", rather than "require that the sequencer is not allowed".
There are other occurences already of the if/revert pattern in our codebase, so IMO readability is sufficient to justify this.

tynes commented on packages/contracts-bedrock/src/libraries/Slot.sol:
```diff
@@ -0,0 +1,69 @@
+        }
+    }
+
+    /// @notice Stores a value in an arbitrary storage slot, `_slot`.
```
Ideally we can get the changes to this contract in https://github.com/ethereum-optimism/optimism/pull/7428

maurelian replied:
Those changes won't be on this branch.
If 7428 does get merged in, will this PR be blocked until the feature branch is rebased?

maurelian committed:
refactor(ctb): Remove revertWhenPaused

maurelian committed:
feat(ctb): Warn on missing new config values.

For values that aren't in pre-existing configs, we'll just warn instead of throwing an error.

maurelian committed:
feat(oco): Just Println for config warning

maurelian commented on packages/contracts-bedrock/.storage-layout:
```diff
@@ -6,26 +6,27 @@
+| failedMessages     | mapping(bytes32 => bool)  | 206  | 0      | 32    | src/L1/L1CrossDomainMessenger.sol:L1CrossDomainMessenger |
+| __gap              | uint256[42]               | 207  | 0      | 1344  | src/L1/L1CrossDomainMessenger.sol:L1CrossDomainMessenger |
+| PORTAL             | contract OptimismPortal   | 249  | 0      | 20    | src/L1/L1CrossDomainMessenger.sol:L1CrossDomainMessenger |
+| superchainConfig   | contract SuperchainConfig | 250  | 0      | 20    | src/L1/L1CrossDomainMessenger.sol:L1CrossDomainMessenger |
```
big diff due to the column width change. The only real diff is the addition of the new superchainConfig var at the end

refcell commented:
FYI - adding the https://github.com/ethereum-optimism/optimism/labels/C-urgent label to test if merge queues still isn't working as intend.

refcell commented:
NVM - base branch isn't develop
